File: src/main/java/com/budgetbuddy/model/SpendingPrediction.java
package com.budgetbuddy.model;

import jakarta.persistence.*;
import java.time.LocalDate;

@Entity
@Table(name = "spending_predictions")
public class SpendingPrediction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "prediction_date", nullable = false)
    private LocalDate predictionDate; // When this prediction was made

    @Column(name = "forecast_start_date", nullable = false)
    private LocalDate forecastStartDate; // Start of forecast period

    @Column(name = "forecast_end_date", nullable = false)
    private LocalDate forecastEndDate; // End of forecast period

    @Column(name = "category")
    private String category;

    @Column(name = "subcategory")
    private String subcategory;

    @Column(name = "predicted_amount", nullable = false)
    private Double predictedAmount;

    @Column(name = "confidence_score")
    private Double confidenceScore; // 0.0 to 1.0

    @Column(name = "prediction_method")
    private String predictionMethod; // "PATTERN_BASED", "TREND_BASED", "HISTORICAL_AVERAGE"

    @Column(name = "risk_level")
    @Enumerated(EnumType.STRING)
    private RiskLevel riskLevel; // LOW, MEDIUM, HIGH

    @Column(name = "is_overspending_risk")
    private Boolean isOverspendingRisk = false;

    @Column(name = "created_at")
    private LocalDate createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDate.now();
    }

    public enum RiskLevel {
        LOW, MEDIUM, HIGH
    }

    // Constructors
    public SpendingPrediction() {}

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public LocalDate getPredictionDate() {
        return predictionDate;
    }

    public void setPredictionDate(LocalDate predictionDate) {
        this.predictionDate = predictionDate;
    }

    public LocalDate getForecastStartDate() {
        return forecastStartDate;
    }

    public void setForecastStartDate(LocalDate forecastStartDate) {
        this.forecastStartDate = forecastStartDate;
    }

    public LocalDate getForecastEndDate() {
        return forecastEndDate;
    }

    public void setForecastEndDate(LocalDate forecastEndDate) {
        this.forecastEndDate = forecastEndDate;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getSubcategory() {
        return subcategory;
    }

    public void setSubcategory(String subcategory) {
        this.subcategory = subcategory;
    }

    public Double getPredictedAmount() {
        return predictedAmount;
    }

    public void setPredictedAmount(Double predictedAmount) {
        this.predictedAmount = predictedAmount;
    }

    public Double getConfidenceScore() {
        return confidenceScore;
    }

    public void setConfidenceScore(Double confidenceScore) {
        this.confidenceScore = confidenceScore;
    }

    public String getPredictionMethod() {
        return predictionMethod;
    }

    public void setPredictionMethod(String predictionMethod) {
        this.predictionMethod = predictionMethod;
    }

    public RiskLevel getRiskLevel() {
        return riskLevel;
    }

    public void setRiskLevel(RiskLevel riskLevel) {
        this.riskLevel = riskLevel;
    }

    public Boolean getIsOverspendingRisk() {
        return isOverspendingRisk;
    }

    public void setIsOverspendingRisk(Boolean isOverspendingRisk) {
        this.isOverspendingRisk = isOverspendingRisk;
    }

    public LocalDate getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDate createdAt) {
        this.createdAt = createdAt;
    }
}

File: mobile-version/app/src/main/java/com/budgetbuddy/mobile/data/model/SpendingPrediction.kt
package com.budgetbuddy.mobile.data.model

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import java.time.LocalDate

@Entity(
    tableName = "spending_predictions",
    foreignKeys = [
        ForeignKey(
            entity = User::class,
            parentColumns = ["id"],
            childColumns = ["userId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index(value = ["userId"]), Index(value = ["forecastStartDate", "forecastEndDate"])]
)
data class SpendingPrediction(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    
    val userId: Long,
    
    val predictionDate: LocalDate, // When this prediction was made
    
    val forecastStartDate: LocalDate, // Start of forecast period
    
    val forecastEndDate: LocalDate, // End of forecast period
    
    val category: String?,
    
    val subcategory: String?,
    
    val predictedAmount: Double,
    
    val confidenceScore: Double?, // 0.0 to 1.0
    
    val predictionMethod: String?, // "PATTERN_BASED", "TREND_BASED", "HISTORICAL_AVERAGE"
    
    val riskLevel: RiskLevel?, // LOW, MEDIUM, HIGH
    
    val isOverspendingRisk: Boolean = false,
    
    val createdAt: LocalDate? = null
) {
    enum class RiskLevel {
        LOW, MEDIUM, HIGH
    }
}

